pkg/tsdb/loki/loki.go

```diff
@@ -188,34 +189,72 @@ func queryData(ctx context.Context, req *backend.QueryDataRequest, dsInfo *datas
 		plog.Error("Failed to prepare request to Loki", "error", err, "duration", time.Since(start), "queriesLength", len(queries), "stage", stagePrepareRequest)
 		return result, err
 	}
-	plog.Info("Prepared request to Loki", "duration", time.Since(start), "queriesLength", len(queries), "stage", stagePrepareRequest)
 
-	for _, query := range queries {
-		ctx, span := tracer.Start(ctx, "datasource.loki.queryData.runQuery")
-		span.SetAttributes("expr", query.Expr, attribute.Key("expr").String(query.Expr))
-		span.SetAttributes("start_unixnano", query.Start, attribute.Key("start_unixnano").Int64(query.Start.UnixNano()))
-		span.SetAttributes("stop_unixnano", query.End, attribute.Key("stop_unixnano").Int64(query.End.UnixNano()))
-
-		if req.GetHTTPHeader("X-Query-Group-Id") != "" {
-			span.SetAttributes("query_group_id", req.GetHTTPHeader("X-Query-Group-Id"), attribute.Key("query_group_id").String(req.GetHTTPHeader("X-Query-Group-Id")))
-		}
-
-		frames, err := runQuery(ctx, api, query, responseOpts)
-
-		queryRes := backend.DataResponse{}
-
-		if err != nil {
-			span.RecordError(err)
-			span.SetStatus(codes.Error, err.Error())
-			queryRes.Error = err
-		} else {
-			queryRes.Frames = frames
+	plog.Info("Prepared request to Loki", "duration", time.Since(start), "queriesLength", len(queries), "stage", stagePrepareRequest, "runInParallel", runInParallel)
+
+	// We are testing running of queries in parallel behind feature flag
+	if runInParallel {
+		resultLock := sync.Mutex{}
+		err = concurrency.ForEachJob(ctx, len(queries), 10, func(ctx context.Context, idx int) error {
+			query := queries[idx]
+			ctx, span := tracer.Start(ctx, "datasource.loki.queryData.runQuery")
+			span.SetAttributes("runInParallel", true, attribute.Key("runInParallel").Bool(true))
+			span.SetAttributes("expr", query.Expr, attribute.Key("expr").String(query.Expr))
+			span.SetAttributes("start_unixnano", query.Start, attribute.Key("start_unixnano").Int64(query.Start.UnixNano()))
+			span.SetAttributes("stop_unixnano", query.End, attribute.Key("stop_unixnano").Int64(query.End.UnixNano()))
```

Comment:
I think there's no `span.End()` for this one, right?

Reply:
Good catch, fixed in https://github.com/grafana/grafana/pull/74064